### ISSUE TITLE
DVI/HDMI-version?

### ISSUE THREAD

**Issue by jbilander** [bitluni/ESP32-S3-VGA#3]
Hi, it would be great if you could try it with a DVI PMOD too,
I made a 16-bit version (RGB-565) of the icebreaker-pmod that should be possible to handsolder.
https://github.com/jbilander/icebreaker-pmod/tree/master/dvi-16bit/v1.2a
Would be cool if you could try it with your code.

More info on the original 1bitsquared PMOD here:
https://1bitsquared.com/collections/fpga/products/pmod-digital-video-interface

Regards, Jorgen

**Comment by jbilander** [comment]
A quick follow-up on the 16-bit DVI-module. yay, It works!, Here I try it with a Tang Nano 9K FPGA-board:
https://drive.google.com/file/d/14sALyBH27fKeR5_NWQFtl_yHK9Rvj7s2/view?usp=sharing
Next up will be to try and drive it from the `ESP32-S3-DevKitC-1-N8R8` devboard. I have followed same pinout as bitluni's VGA pin-mapping, but I still need to assign the DE and Pixel-CLK signals. Thinking of either GPIO 47,48 or GPIO 19,20 for this purpose:

![IMG_8506](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/a3387c2b-769d-4ca6-9371-568846e832ba)
![IMG_8492](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/da487a06-f303-4849-b2a8-4337e5d70fd8)

@bitluni Let me know if you want a blank PCB of the 16-bit DVI-module to build and tinker with. I have some spares, I'll send one to you free of charge, in that case drop me an email jorgen.bilander at gmail com

Regards, J

**Comment by sobieh** [comment]
I never attempted to use any digital interface like dvi, hdmi with mcu so i would be happy to see some working example :)

**Comment by jbilander** [comment]
@sobieh 
Hi, it's direct output from ESP32, no CPLD involved. I used a simple veroboard for prototyping, pic above. Yes indeed it'll be interesting to see if the ESP32-S3 can drive this DVI-module, hopefully it can. I still have left to assign and solder the two wires DE and PCLK. I will put in a 33 Ohm damping resistor on the PCLK-line too when I do this, just like I did with the FPGA-prototype. From what I can see in the ESP32-S3 datasheet any free GPIO can be used with the LCD interface:

![image](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/4d6c2fac-9aa8-410c-b70c-5737325f1c4e)

Hopefully I will have time to finish up soldering today and perhaps start writing some code. If you have a code example for this please let me know.

**Comment by BrainBoardz** [comment]
Neat hardware! I would avoid assigning GPIO 19 or 20 for anything if at all possible. Those are typically reserved for the USB +/- pins! GPIO 47/48 should be fine.  

 I did have some success with my prototype S3-VGA (it worked, yay!) board and a very cheap VGA to HDMI converter. These are  inexpensive converters, so potentially a viable option given that VGA is now less common than HDMI. 

It's a bit unfortunate that the vast majority of resolutions in the library are 4:3. I'd like to see how the HDMI converter would work with  a widescreen (16:9) resolution. Ideally something like 854x480 (Full Wide Video Graphics Array) or similar. Perhaps a lower resolution one for gaming as well . I think that 854x480 would provide a very useable resolution and it would not stress the S3 to much. This assumes that it can be generated of course!  I'm reading over the project files to see if I can figure out how to calculate the proper timings etc. Having some guidance on calculating new modes would be very useful.

**Comment by bitluni** [comment]
This is an amazing work!
@jbilander thanks for the offer. the risk very high that it ends up on the pile in lack of time :-(
I even have the HDMI ICs that are used on the board here. My plan was to use those with the old esp32 vga implementation. still didn't happen.

**Comment by jbilander** [comment]
Hi, finally I had time to look at this and wow it works like a charm, video below:
https://drive.google.com/file/d/1HP_4MCiUtJIRtubJrGJjzsNwiGtcmvMG/view?usp=sharing
Running 640x480 there. I haven't tried going higher yet.

![IMG_8812](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/b8e178b1-52f8-4cff-89cb-c9539a718ee0)

I used GPIO 47,48 for the PCLK and DE.

Example code is here:
https://github.com/jbilander/Exploring-Graphics/blob/main/ESP32-S3/main/Main.cpp

I had to config the 8MB Octal SPI PSRAM to run at 120 MHz, I enabled the experimental feature in menuconfig (running idf.py menuconfig and I was able to set it at 120 MHz). The frame buffer runs as a double buffer.

Someone should code a game like flappy bird or something and try some gaming on the ESP32-S3. That could be fun. I think maybe a USB gamepad or joystick could be connected to the unused USB-jack (GPIO 19,20). The ESP32-S3 has USB OTG 1.1 built-in so could potentially act as a usb-host-controller for a gamepad, me thinks.

**Comment by jbilander** [comment]
Hi, @BrainBoardz 
I tried 854 x 480 resolution. I used timings I found on the RPi forum:

hdmi_timings=854 0 26 80 104 480 1 3 2 15 0 0 0 60 0 31750000 3

from here:
https://forums.raspberrypi.com/viewtopic.php?t=269968

hdmi_timings string seems to be of this format:
hdmi_timings=<h_active_pixels> <h_sync_polarity> <h_front_porch> <h_sync_pulse> <h_back_porch> <v_active_lines> <v_sync_polarity> <v_front_porch> <v_sync_pulse> <v_back_porch> <v_sync_offset_a> <v_sync_offset_b> <pixel_rep> <frame_rate> <interlaced> <pixel_freq> <aspect_ratio>

hence:

h_front_porch = 26
h_sync_pulse = 80
h_back_porch = 104
v_front_porch = 3
v_sync_pulse = 2
v_back_porch = 15

The monitor I tried this on syncs and shows a pic but I don't think the double frame buffer in the ESP32-S3 module is quick enough, I get garbled pic but it fills the wide-screen correctly. 

![IMG_8841](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/5feb086f-2d33-4afa-8ff2-abdbfde4e0cf)
![IMG_8843](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/593e614a-f25e-48da-ae93-923804482c45)

I get the same garbled output on 640x480 if I set the PSRAM speed back to 80 MHz so I guess that's the culprit here them being too slow.

I also tried 852 x 480, same thing. Perhaps if I change to single buffer or bounce buffer or something, but among the timing formats, only one format, 640 × 480p @ 60 Hz, is mandatory and common for both VESA and CTA-861 standards. Both PC Monitors and TVs have to support this particular mode, so I think I will go with that if I'll try and develop a game. 4:3 aspect ratio is also very retro which I like. Regards, J

**Comment by BrainBoardz** [comment]
Hi Jorgen:

Thanks for the details on the video mode. That's a big help.  I didn't
consider that this mode could be more than the ESP32S3 could handle.  I
think I am going to bash away and see if I can find a (any!) working 16:9
mode. I do see the advantages of 640X480 for retro gaming in particular,
but it is very limiting to have only a selection of 4:3 modes on offer
given the prevalence of 16:9 screens.

bye for now,




*Jeremy Littler*

*LEAD Emerging Technologies*

The Creative School

Toronto Metropolitan University

416-979-5000 ext. 552222

RCC 353

**Comment by jbilander** [comment]
@BrainBoardz 

No problem, I'm glad if it helped. BTW I found this online calculator where you can punch in your resolution and refresh rate and you get a lot of details regarding front porch, back porch values and so on...

https://tomverbeure.github.io/video_timings_calculator

BTW, I tried 640 x 400 too, which is aspect ratio 16:10 (not same as 16:9 though), and it looked nice on the monitor I tried on which has native resolution 1680 x 1050 hence also 16:10 aspect ratio, however the monitor didn't autoswitch to widescreen mode which I think was a bit weird. I had to manually go into the display setting and set it to wide, pic below is after that setting was done. I don't know how useful that resolution might be for you and how compatible it is?, but I though I should mention it at least, that it worked in my testing.

![IMG_8845](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/8ca07243-2319-4684-9750-b5aedf990fbb)

Regards, J

**Comment by BrainBoardz** [comment]
Hi Jorgen:

Thanks. I did play with that site at one point. It's extremely useful. The
600 X 400 mode is absolutely fine by me. it's not 16:9 but the ratio is
close enough to widescreen. I would expect it to be less challenging for
the S3 as well. That's a bit weird about having to switch to widescreen
mode. I'll play with this and see if it is a common issue with other
monitors. I always hook-up my VGA test boards to a VGA to HDMI converter so
I've not seen that issue. So perhaps this won't be a problem if people are
using converters. If you could send me the timings you used for the 640X400
mode I would greatly appreciate it.  If I can get it working with the S3 I
will submit a request to bitluni on Github to add it to the list of
supported modes.

bye for now,


*Jeremy Littler*

*LEAD Emerging Technologies*

The Creative School

Toronto Metropolitan University

416-979-5000 ext. 552222

RCC 353

**Comment by jbilander** [comment]
@BrainBoardz 

Hi, no problem, I used the values from link below:
http://tinyvga.com/vga-timing

This one to be exact:
http://tinyvga.com/vga-timing/640x400@70Hz

Regards, J

**Comment by BrainBoardz** [comment]
Awesome! Thanks again.


*Jeremy Littler*

*LEAD Emerging Technologies*

The Creative School

Toronto Metropolitan University

416-979-5000 ext. 552222

RCC 353

**Comment by jbilander** [comment]
@BrainBoardz 
Hi, I started implementing flappy bird on the ESP32 just to see if it could work.
https://drive.google.com/file/d/1_6dJlVb1nWwBIv3ym8lq7n9sIrbvJ14P/view?usp=sharing
As you can see it struggles a bit when there are a lot of pipes (bitmaps) to move on screen.
I'm using lvgl so perhaps it's not optimal for this task (running games) but kind of cool it works anyway.
Regards, J

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

That's really neat! I spent some time over the holiday thinking about what it might take to make a really viable ESP32S3 system for gaming and other applications. This is obviously a very very ambitious wish-list and just some potential directions this could go.     

1) I saw an example of a RP2040 based board that utilizes a Lattice 
SII9022ACNU HDMI Transmitter. Basically a RGB to HDMI chip. Not sure if the ESP32S3 could work with this (maybe the PIOs in the RP2040 are the secret sauce that makes this work), but anything that enables the use of HDMI would be a big step forward in my book! VGA output is neat but a "DVI+AUDIO" solution would be...mega cool! 

 I had started to look at some Chinese options for VGA to HDMI these but the documentation and sourcing of these chips is very murky. The Lattice IC is actually pretty inexpensive and it looks like it is commonly available for around $3.50 USD.  

2) I was able to get line level audio output from a prototype ESP32S3 VGA board by implementing a custom I2S solution using the PCM5102APWR. I used GPIO 45, 46 and 48 for the I2S.  This worked amazingly well! I had VGA and really clean sounding audio from a MicroSD card, albeit it worked only at the lower VGA resolutions. I does look like the SII9022ACNU can also work with I2S. So perhaps the 5102 chip would be redundant if a Lattice chip was used. That would lower the overall BOM cost.  

3) I've been investigating the use of an SNES controller (pins 43,44,48). I did look at other options (USB perhaps), but this appears to be the easiest joystick method to implement. However, this leads me to another issue. As implemented (with Octal PSRAM S3's)  Bitluni's VGA library uses up the vast majority of the available pins. Looks like I just have enough for one SNES controller, but I had to give up the serial pins (ok for me as I am using USB for the programming anyway).  Very few pins are left (well other than off-limits strapping pins). I think I can implement the SNES joystick with the S3 (going to test this will a board) but that pretty much eliminates any other I/O capabilities. I'm wondering if the VGA pins could be slightly scaled back to free up just a few more pins.  Maybe a dual SNES capability or just enough for an SNES and an I2C sensor. It would enable lots of applications.  

4) It would be amazing if an approachable gaming library could be developed for this. Something based on Circuit Python perhaps.  That's one heck of a reach, but I would love to see what could be developed using Python or Circuit Python. Even Arduino perhaps. Somewhat like FabGL? It looks like bitfixer has been able to squeeze a lot of performance out of the S3, but he is using the ESP-IDF to accomplish that and I understand he is not using LVGL. That method is not exactly easily approachable for novice programmers. I think Bitluni's implementation of the Adafruit's GFX library really demonstrates up what is possible (performance notwithstanding). A complete library with sprites etc. would take this to the next level. Looks like it is either relatively easy to use but slow or faster but more complicated. Maybe the LVGL approach could be tweaked, or even a library developed that is more optimized.  

So lots of crazy ideas, but I think if we want a true ESP32S3 gaming system this could be some directions to look at.

**Comment by jbilander** [comment]
Hi @BrainBoardz 

Nice ideas. I had a similar idea to create a carrier board PCB with a ADV7513 HDMI tx chip (LCSC has them in stock) and put a usb-hub chip on board together with many female USB connectors so that several usb-gamepads can be used simultanously. It would require usb-hub functionality though which in turn requires development from the ESP-team I think, since stuff is limited what you can do yourself in that regard. I implemented usb-mouse functionality from their example code but currently only one device supported. I have also tested audio and wav-playback which works fine but I think mp3-playback would be the preferred choice since file size is a concern unless you make a SDcard a mandatory requirement. I thought of converting MP3 files to C Header files with xxd and perhaps put a PCM5102A chip on board if analog external line out is going to be needed. There is a nice github repo here https://github.com/schreibfaul1/ESP32-audioI2S that can do mp3 playback. I haven't dived into it though.

I have now put my hacking in a public github repository if someone wants to take a closer look. 

https://github.com/jbilander/ESP_Control_Deck/

Flappy Bird is not fully implemented and perhaps won't be. It was just a test, but maybe I will try to get HAGL or FabGL up and running with this example instead of LVGL and see if better performance can be achieved. I think this will require some work on a backend implementation though so might be tricky.

I don't fully understand how the two framebuffers work on the esp_lcd_rgb_panel, as you can see here i tried allocating ram from MALLOC_CAP_SPIRAM but this slows down things.
https://github.com/jbilander/ESP_Control_Deck/blob/main/ESP32-S3/main/Main.cpp#L368
and you also have a setting here:
https://github.com/jbilander/ESP_Control_Deck/blob/main/ESP32-S3/main/Main.cpp#L169
Don't know what this really does and how the buffers are allocated if only void *buf1, *buf2 is used with this setting. Perhaps it is allocating from SRAM.

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

That ADV chip is interesting and reasonably priced. I'm going to look at the technical docs on that! In terms of the USB that would be amazing to have, but I it does appear to be challenging to implement as you have indicated.  I  even looked at using (via Bitluni's work) a CH559 for that purpose, but programming that chip looks quite complex and it would make the whole thing dependent on one supplier afaik and require more passives + a more costly board. Might be easier to just use a SM32 to expand the IO capabilities! A absolutely bare bones Blue Pill type implementation ha ha.  

I think that PS/2 ports might work in in similar fashion to how those operate on the FabGL boards. Those are probably easier to implement. They are old school, but if they worked that would be one option.  But, as it stands there are not enough free pins on an Octal ESP32S3 to have a DAC, SNES and PS/2 ports (well maybe with a switch between PS2 mode and SNES mode!). I think what would really help is if Bitluni's library provide an easily selectable mode that enabled VGA with fewer pins. That looks possible to modify in Bitluni's code and Bitfixer also did help me with some info on the pins count for VGA. That said, it would be nice if the pin assignments for VGA on the ESP32S3 could be standardized across multiple libraries as that would be very hard for end users to adjust. An ESP32S3 VGA consortium!   Having divergent pins and divergent VGA libraries is a roadblock to mass adoption IMO. I think that we would want a standard FabGL style ecosystem for this to be really useful and flexible. Maybe it too early to expect that level of standardization but I can see the potential if we had it.  

I was able to get MP3 playback on an SD card by injecting MP3 playback code in Bitluni's Adafruit GFX example. It was hacky but it worked well. I think that we are going to have to work on the assumption that any useful VGA board would have SD hardware. It's pretty much essential for even MP3 audio and also for other assets.  So perhaps a reference design that standardizes this as well.  

I'm quite terrible at coding, so I hope you figure out the performance issues. It's way over my head.  The progress you made with the game is very impressive and it points the way to what could be possible with the S3.

**Comment by jbilander** [comment]
Hi @BrainBoardz 

Yes I agree, ADV7513 is interesting to use, and they have good documentation about how to do the hw-design with it. I'm actually at the moment designing for it in another project so hopefully I will have some experience with it in the near future. I don't know if the I2C also has to be hooked up (to control HDMI settings) but if that is the case you're looking at another two GPIO pins occupied. This means you pretty much have to get USB-Hub working in order to have more than one USB gamepad working. Especially if you have an SD-card slot on-board too which takes another three GPIOs in 1-bit SDIO mode. If you follow the links below you can see USB-hub seems to be on the todo list. Fingers crossed is not too far away. I guess one could try and implement it oneself (according to the last comment) but it's a bit over my head and knowledge about the USB host stack at the moment. It would be enough to get it to work with HID-devices (mouse, keyboard and gamepad) but still not trivial to implement for me.

* No Hub support (currently only supports a single device)

https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/usb_host/usb_host_notes_index.html

https://esp32.com/viewtopic.php?t=34167#p125175

Also regarding other graphic libraries and performance. I stumbled upon this repository and if you click the image you can see a short youtube video, he is using @bitluni's ESP32Lib, looks to be very good performance.
https://github.com/moefh/esp32-loser/

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
I'd love to hear how the ADV7513 works out!  Perhaps Bluetooth controllers instead? I've see examples of PS controllers being used in this role with the older ESP32s I did actually test this approach a bit, but it did not work at the time. I think I was not implementing it correctly at the time. Perhaps it would be good to give it another look. Having the SD support might be handy for storing device id's or even to identify what type of controllers are connected. 

It certainly looks like GPIOs are a big issue. Particularly if I2C is required for HDMI (yikes!). I'm wondering if the 19/20 USB pins could be made to do double duty between programming mode and external device mode. Perhaps using a switch to change the routing to say a USB-A connection. 

That library is interesting. I do like the greatly reduced GPIO usage. I'm not sure that we need 720P or high bid depth resolutions to create fun games, and the lower overhead could certainly help with performance.

**Comment by jbilander** [comment]
Hi @BrainBoardz

Using PS3-controllers might be a good alternative since they are very good quality, although the BT-pairing might be a small hurdle for plug-and-play, there is this library that looks to do the job well, I haven't tried it though. Could be fun to explore if you have the spare time:
https://github.com/jvpernis/esp32-ps3

Regarding USB-Hub on the ESP32-S3. It sounds like even when usb-hub is fully supported it'll effectively be max two active HID-devices working at the same time because of limitation in number of channels (8).
The upcoming ESP32-P4 will have 16 channels going by the forum post here:
https://esp32.com/viewtopic.php?p=125441#p125441
and it will have some 50 programmable GPIOs and 768KB of on-chip SRAM so perhaps a better MCU to use for a ESP games console type of board. However no embedded WiFi or BT on that one.

If you are interested in ADV7513, here are some useful links I've come across:
https://www.analog.com/media/en/technical-documentation/user-guides/ADV7513_Hardware_User_Guide.pdf

Especially page 40 and 45 are helpful when designing and doing PCB layout.

As you see the ADV7513 requires a 1.8V LDO, I think I'm going to use the same type they use on the board below, MCP1702T, page 10 on the schematic:
https://www.96boards.org/documentation/consumer/avenger96/hardware-docs/files/avenger96-schematics.pdf

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

I'll check that library and re-visit the PS2 controller implementation on the weekend, and see if I have better luck this time around. The ESP32-P4 looks promising. I think the issue might be the cost as compared to the S3. I guess we will have to see if that as Espressif are very cagey about their release dates. Sill, it would be nice to get the DAC (possibly used for video timing?) and Ethernet capabilities back.  I did take a look at the docs for the ADV and I saw that is does indeed need 1.8V. Not a major show stopper probably. There's loads of diodes and caps though! I see they are using diode arrays.  

 I'd be really interesting in any progress you make on implementing this. I guess we would still need to implement the VGA resistor ladder?

 If the PS2 Bluetooth approach worked then theoretically we could create an small HDMI unit that connects to a TV directly (powered by the USB port on the TV?) and utilizing a Bluetooth PS2 controller. I was able to power my test VGA board using the TV USB port. Keeps it tidy. It could be a cool little ultraportable gaming machine.

**Comment by jbilander** [comment]
Hi @BrainBoardz , the resistor ladder is not needed unless you plan on having an analog VGA output. Personally I don't need VGA, it's kind of obsolete if you have HDMI. If I were to design for VGA I would prefer to use ADV7123 or ADV7125 but a resistor ladder can be a good enough (poor man's) DAC and certainly a cheaper way to do it. In that case I think it's important to use resistors with high accuracy for the ladder to get reliable consistent output voltage values (=colors).

In regards to powering from the TV's USB jack, it depends on the current consumption. Regular USB I think is specified to be able to supply 500 mA on VBUS (+5V). If the ESP Board draws more than that with everything running I don't know, it depends on how much stuff is on the board in terms of driving peripherals and how hard the MCU has to work, number of GPIOs used, CLK-frequency etc. Nowadays there's USB-C and that's a completely different story when it comes to powering, it uses negotiation communication between devices to decide capacity and what side is acting as supply. It's a bit more work to design for USB-C. You can read more here: 
https://dubiouscreations.com/2021/04/06/designing-with-usb-c-lessons-learned/

I let you know how the ADV7513 endeavor goes.

Great if you can get the PlayStation controllers working. I guess you mean PS3-controllers because aren't PS2-controllers wire-only and no built-in Bluetooth? That's how I remember it.

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

Thanks for the clarification. I agree that VGA is ultimately a bit of a dead end. HDMI all the way I say. If the VGA ladder elements can be skipped completely in a design that is even better. I'm a bit confused about how the ESP32S3 would be connected to an ADV7513. I see mention in their documentation of Syncs and Clocks and some color formats. So, this is something that has to programmed (e.g., it's not based on a specific protocol or a format conversion)?  If this turns out to be actually possible with an ESP32S3 then that would be a wow moment!  

Sorry, yes PS3. I've grabbed mine and will be testing forthwith. I'll let you know how that goes.

I was able to get my VGA board (+ the DAC) running off a USB 2 TV connector. I think 500mA should be enough, but that is not verified with a fully equipped board (say with a ADV7513 and other bits). I've read the S3 consumes about 70mA base and 120mA if wi-fi is enabled. Still, with other peripherals this could become an issue, though 500mA is a fair bit of headroom to play with. My boards are all utilizing USB-C these days, so plenty of mA with that. Having the capability of powering off a TV USB port would be a very nice capability. I'm thinking specifically of TVs being used for signage, as these may not have readily available AC power located near to the TV. It could reduce cable clutter and the need for wall warts.

**Comment by jbilander** [comment]
Hi, @BrainBoardz Regarding how to connect the ESP to the ADV7513, it's no different from how my DVI-module with the SII164-chip is currently hooked up to the ESP.

https://github.com/jbilander/ESP_Control_Deck/blob/main/ESP32-S3/main/Main.h#L20

You assign any free GPIO pins for PCLK, HSYNC, VSYNC, DE and the 16 RGB_DATA pins on the ESP and route them to the corresponding input pins on the ADV7513. Unused RGB Data pins are grounded. In addition to that you add/route the three I2S audio pins and the two I2C-pins (SDA, SCL). I2C would be for having the possibility to control HDMI-settings like de-skew and color formats etc, I think this will not be needed since RGB is by default but it could be nice to have the possibility to do settings this way if there are free GPIO pins on the ESP. For the PCLK trace it's good to have a 33 Ohm series resistor if the trace length is longer than an inch on the PCB (rule of thumb) to avoid signal reflections.

Regarding the 500 mA limitation, you also have to consider the HDMI-chip draw and the HDMI requirement to supply +5V on pin 18. I don't know exactly what the specification says. According to this Ti-document it seems to be 55 mA:
https://www.ti.com/lit/an/snva412c/snva412c.pdf?ts=1706804074777

Yes, It would be nice with a compact design that can be hooked up like a chromecast thingie to the TV.

I noticed youtuber [DroneBot Workshop] has a video about _Using PS3 Controllers with ESP32_. I think he's using the library we talked about before. I haven't watched the video yet.

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

Thanks for the info. Looks like there isn't going to be enough free pins to do pull this off  (particularly on the Octal PSRAM equipped ESPs). I count at least 25 for just the video (4 + 16 + 2 if my math is correct). Not sure about whether the ground pins are additional to this. Excluding strapping pins, there appears to be 28 (-1 perhaps for an LED) available.  That would leave at most 3 pins for other purposes. Not even enough to enable an SPI connection to say an STM32 (to implement a controller or perhaps SD card).  I have been able to use strapping pin 45 without locking up the ESP, but Pins 46 and 3 are very intolerant of being used for anything in my experience. Maybe my counts are off, but it certainly looks the ESP P4 is the only way to go. 

 I did re-test and was hopeful that a PS3 controller would work, but I don't have a genuine Sony unit. Tried all sort of tools like SCP Toolkit etc. But, my clone is not working properly and I cant see it at all in sisaxis. I'm chasing an original unit down, but those are no longer easy to get in retail form. Probably not a good direction given that the OEM units are Ebay items and even used ones are pricey on Amazon.

**Comment by jbilander** [comment]
Hi @BrainBoardz 

These are the pin-assignments I'm currently using and have tested working ok. RGB data pins are the same as bitluni's:
![ESP32-S3_DevKitC-1_pinlayout_v1 1](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/64373205-30fe-4886-a244-3a3385694ecd)

Perhaps the strapping pins GPIO 0,3,46 can be used for I2C?. The three pins INT, SDA, SCL has to be designed with pull-up resistors which might be bad for bootstrapping? I don't know. I forgot we probably want the Interrupt signal connected too.

![image](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/01284be5-4eb2-43c3-b1c2-79e3c4a041d6)
![image](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/caf7d78d-7a5e-4f78-b49e-14d363c5938c)

Regarding PS3-controllers yes the clones can be a hassle to get working. I remember a friend had big problems to get em working with a RetroPie installation, while genuine ones were no problem.

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

I purchased a clone PS4 controller on the weekend.  I just tested it with Sixaxis and it worked perfectly! This cost me about $20.  Way less expensive than OEM PS3 or PS4 units. So definitely a promising contender! Looks like it should work fine for ESPs as there are libraries for this (that is TBD though). I'm going try out the controller out and report back. If it does I will order a different brand and see if this the PS4 clones are consistently useable. Looks like the PS4 units are more compatible to the OEM models.   

Thanks for the pinout info. No wasted space there eh. Maybe I am missing something, but I see only 3 pins for SPI. 

Just a thought, but I'm wondering if GPIO43 or GPIO44 could be repurposed. Assuming that bluetooth is confirmed as a viable controller method, then as long as USB (19/20) pins are available those could be used exclusively for the ESP programming. My boards don't have  even have a UART. That could free up 2 pins, and we might even be able to bring back the LED (ha ha).

**Comment by BrainBoardz** [comment]
Bummer. It turns out the these controllers require Bluetooth Classic. This mode is not supported on the S3. The S3 only supports BLE and that is not cross compatible with Classic BT. This library incompatibility issue has been a common theme with the S3. All of the standard PS4 libraries I tried crashed with errors. So back to square one. The only alternative I can see at this point would be using a compatible BLE controller. There is a library for BLE (that does not crash with the S3. I did actually verify that):

https://github.com/lemmingDev/ESP32-BLE-Gamepad

This seems more focused on turning the S3 into a BLE controller, but maybe the other direction?  I've posted a question on their forum to see if it can work in reverse and if there are any recommended controllers (8bitdo possibly). Failing that I think we are out of luck.

**Comment by BrainBoardz** [comment]
BluePad32+ESP does offer support for the ESP32S3. But, the list of BLE compatible controllers is very limited. Stadia (defunct), Steam (discontinued), Xbox 1914 (pricey and no idea if the clones would work).

https://github.com/ricardoquesada/bluepad32/blob/main/docs/supported_gamepads.md

Looks like SNES (or perhaps USB?) is the only controller option, extra-GPIO pin usage notwithstanding. Maybe a simple BLE controller based on an ESP32S3! 

https://www.aliexpress.us/item/3256804105847572.html?gatewayAdapt=glo2usa4itemAdapt

**Comment by jbilander** [comment]
Ok, bummer. Maybe best is to wait and see what the ESP32-P4 has in store for us then. Great that you took the time to test this though, thanks for that. Regarding the SD-card, I'm using 1-bit SDIO-mode, tested ok, so only three pins required compared to SPIs four. This saves one GPIO but still it seems we are a bit short on GPIOs. From a performance perspective it would be best to run the SD-card in 4-bit mode absolutely, but I guess the "crippled" 1-bit mode would be good enough. GPIO43 or GPIO44 could probably be repurposed but the downside not having the USB-pins available for a USB-Mouse is significant in my opinion. I think it would be nice to have the possibility to control some type of games with a Mouse, like strategy games or old school point n' click adventures should anyone feel like porting or making one.

Regards,
Jorgen

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

Indeed! The P4 is looking promising. Loads of GPIOs on that beast. I think the issue with the P4 is that it was announced a while back and there has been no release date announced.  One post mentioned possible production issues with it.  Hopefully that will get sorted out.   

I keep running into limitations with the S4. No DAC, no Ethernet and now  no BT Classic. I wish Espressif had just added USB support to the older ESP32 chip, some more pins and performance/memory tweaks. Not sure if they could have done that from a technical standpoint, but they certainly threw out a lot of the essential ESP capabilities with their transition to S4.   

I wasn't even aware 1-bit mode was a thing.  Agreed about the USB mouse support being important. I think it could be  something that could (maybe) be switched (between mouse or programming over USB mode perhaps), but that would be hacky.  I guess we will have to wait an see what develops. It might even come in the form of a updated RP2040 module or perhaps a low cost SOC chip with built-in HDMI capabilities.

**Comment by BrainBoardz** [comment]
Hi Jorgen: 

Sorry, I can't fix the Mouse requirements, but it does look like there are some Bluetooth controllers that might work with BluePad32. The developer got back to me about controllers. One of recommended ones s the Terios T3. I'm actually hoping the GameSir TS3 will also work as it is on Amazon. It appears that if the controller supports Android devices this could work with the S3. That would be amazing. I will test this out and report back. Going to need to get a bank loan for controllers!

**Comment by jbilander** [comment]
Hi @BrainBoardz 

According to this post on Reddit the P4 will hit MP in Q3 this year. I guess MP means Mass Production.
https://www.reddit.com/r/esp32/comments/18ui8rq/comment/kiu8z8d/
The price point seems to be in the same range as the S3 which is good news, so it looks promising. Yeah, don't break the bank on BT-controllers, it has to be a cheap solution anyway in my opinion. I guess with the P4 it will be possible to add a BT-dongle/WiFi-module relatively easy on the same PCB should there be a need for that.

**Comment by BrainBoardz** [comment]
That timeline is good to hear. It will be interesting to see what this chip is capable of doing. The extra IO can't hurt and it is great to see Ethernet is coming back! I read the post and saw that they indicated combining it with a  C series microcontroller to get BT and Wifi. That would be a good combination potentially.  The BLE controllers I am looking at aren't very expensive (sub $30). So, I am going to try one out to determine if these are an option for my VGA project.  I'll update this thread if I get positive results.

**Comment by BrainBoardz** [comment]
I was able to get the S3 working with a gaming controller. I used BluePad32 with a Terios T3 Gamepad. Some out of production units might also work. The T3 is not a great controller, but it does work well with the S3. BTW, I have not been able to get any other models of controllers to work. It looks like it is Terios T3 only (other than a handful of out of production units) is the only current option. I will update this thread if I find any other controller that works.

**Comment by jbilander** [comment]
> I'd love to hear how the ADV7513 works out!

Hi @BrainBoardz 

I now have a ADV7513 chip up and running with a Gowin FPGA on a PCB I've designed. This is a flicker-fixer/scandoubler card I'm currently developing for the old Amiga computers (A500/A1200). Still a lot to do in regards to developing firmware for it. Anyway it would be possible to design a PCB and use a microcontroller instead of a FPGA, for example the upcoming ESP32-P4. One thing I've learned is that it's crucial to have the I2C routed and connected since you have to write to different registers on the ADV in order to power up the chip and configure some setup-stuff that is required.

Here is a pic of my current setup with the card. 

![IMG_9470](https://github.com/bitluni/ESP32-S3-VGA/assets/1673918/2e41494a-1dc6-4635-938f-0add1fc1966f)

I currently have a ESP32-devkit board connected to a I2C-pinheader I bodged in only to simplify tinkering and bringing up the ADV. I will later have the I2C-master in Verilog instead residing inside the FPGA.

video here:
https://drive.google.com/file/d/1cCXUOstm10DY3RTalntEts_MJLnChQ2S/view?usp=sharing

**Comment by BrainBoardz** [comment]
That's an impressive achievement!  Good to hear that the P4  might be good enough for HDMI without an FPGA. It will be interesting to see if someone releases a dev board with that functionality.

**Comment by NimbleMotorsports** [comment]
I'd like hdmi output as it is hard to find vga displays anymore, although I just picked up two and including a 32in dell display, and it works great to show my huge 4 digit timer digits.  I can just use a pi zero or similiar,while not expensive, they are still 2 or 3 times more than a esp32.  A usb to hdmi adapter seems like it would do it with the s3, but then you need to know how to make the driver, the suppliers don't seem willing to provide it.
I had another idea of interfacing to the pcie bus...there are graphics cards really cheap now and they have all the memory needed for the display.  it looks like the bus is 100mhz, can't the esp32 overclock to 100mhz?

**Comment by jbilander** [comment]
@NimbleMotorsports 

The ESP32-S3 can run with a clock speed of 240 MHz but I think the problem is the GPIO toggle speed isn't that great. You can read more here how it can be improved: 
https://ctrlsrc.io/posts/2023/gpio-speed-esp32c3-esp32c6/

if you need a monitor with VGA-input that also takes a 15 kHz signal I can recommend the AOC 24E1Q.
It is not that big though, 23.8" - 1920 x 1080 Full HD but a real bargain since it's so cheap for all the inputs it has.

Edit: updated link to pic
[IMG_8512](https://drive.google.com/file/d/1s_Z11SugOOUuyyCl_5AZcrSgwnoWyKyD/view?usp=sharing)

A usb to hdmi dongle would require very high speed usb, The ESP32-S3 only supports USB 1.1 full speed mode, and even if it had high speed capability writing the drivers for it is a big hurdle. Much easier route is to use the S3's (or the upcoming ESP32-P4's) RGB LCD controller 

https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32/api-reference/peripherals/lcd.html#lcd 

with regular GPIOs connected to drive a HDMI-transmitter chip like ADV7513. That is what I did in my experiments shown above.

**Comment by NimbleMotorsports** [comment]
How much is the hdmi chip?  A pi zero is $15 and has hdmi 512mb of ram.  
If the esp32 can do pcie and talk to a gpu board, they are selling cheap on ebay, just seen qty 10 of them for $30, so $3 each typical 1gb of memory.  I'm sure its complicated to do, and would use lots of power, which is nice about the esp32 vs other boards.   The great thing with vga is its just 6 pins (I only need 4 colors myself, I hacked the old driver to support 2bits and got 900x600 4 color)

**Comment by jbilander** [comment]
@NimbleMotorsports 

> How much is the hdmi chip? 

around $3.60 a pop at LCSC:
https://www.lcsc.com/product-detail/Video-Interface-ICs_Analog-Devices-ADV7513BSWZ_C408901.html

> The great thing with vga is its just 6 pins (I only need 4 colors myself, I hacked the old driver to support 2bits and got 900x600 4 color)

ok, if your requirement is only to display some sensor values on a bigger screen and no audio yeah I wouldn't bother with HDMI for that, VGA is working fine and is cheap. You can always find used monitor/TVs that take VGA. Adding a HDMI chip is more if you need to use at least 16-bit RGB colors and also digital audio included in the same cable. My experiments was with a games console in mind based on a cheap ESP32-S3 or the upcoming ESP32-P4.

**Comment by NimbleMotorsports** [comment]
That chip is quite cheap, looks promising!    I have r/c car racing business, and I need outdoor displays for race timings and scoreboards, these usually done via leds panels.   bigger displays are better to be visible in daylight, those are usually hdmi tv's.

**Comment by jbilander** [comment]
@NimbleMotorsports 

I see, so to me it sounds that DVI would be sufficient in your case. You want to be able to plug in a regular hdmi-cable in both ends but doesn't have the need for audio. There is a good and cheap DVI-chip that is suitable for that and it doesn't require you to configure it through i2c-communication in order to start up and start display a pic. Also it only needs a single 3.3V supply compared to the ADV7513 which requires a 1.8V LDO as well. You can of course use HDMI receptacle connectors with the DVI-chip as the signals are the same except no audio is possible to mix in.

https://www.digikey.com/en/products/detail/lattice-semiconductor-corporation/SII164CTG64/6596652

I made a 16-bit PMOD with this chip here and that is what I tested the ESP32-S3 with shown in posts earlier in this thread.
https://github.com/jbilander/icebreaker-pmod/tree/master/dvi-16bit

it is based on the 1-bit squared open source 12-bit PMOD available here:
https://1bitsquared.com/collections/fpga/products/pmod-digital-video-interface

With all that said. A better solution for you might be to use a RPi Pico W and overclock it and do this:
https://github.com/Wren6991/PicoDVI

I'm not sure if that resolution is sufficient in your case. If so you can buy a DVI-sock that is available for the Pico and you solder it directly to the Pico's castellated holes and start hack from there:
https://www.electrokit.com/dvi-sock-for-raspberry-pi-pico

**Comment by NimbleMotorsports** [comment]
I had looked at the pico dvi and thought why can't the esp32 do it also?  I need wifi to send the data to display.
I did find a chip that does a pcie interface with simple parallel bit input, 
https://www.wch-ic.com/products/CH368.html   just looking at it, it seems great to connect it to a pcie gpu board, which has all the memory and processing power compared to a pico or even esp32. 
It is also a general solution that could be used to extend a micro with many other functions.

**Comment by jbilander** [comment]
@NimbleMotorsports 

> I had looked at the pico dvi and thought why can't the esp32 do it also?

The RP2040 MCU on the Pico has a built-in PIO state machine that can run in parallel with the CPU driving GPIO almost regardless how bogged down the central processing cores are. 

https://circuitcellar.com/resources/quickbits/rp2040-programmable-io/

This feature is utilised in the DVI-example but they still have to overclock the chip significantly to keep up. Digital video requires a lot of data throughput and it's very impressive that a small MCU like a RP2040 can pull this off at all. Even though it's only the lower resolutions like 640x480p and perhaps 720p. The ESP32-S3 chip doesn't come with such PIO-feature. Problem is it can't drive the GPIOs quick enough to drive the DVI diff-pair signals without the help of any transmitter chip. The TMDS-channels typically has to be 10x of the pixelclock, so if the pixel clock is 50 MHz the TMDS signalling has to be around 500 MHz, hence we are talking GHz-frequencies for very high resolutions. This TMDS encoding is what a DVI/HDMI chip does for you. Maybe the upcoming ESP32-P4 will be better in regards to GPIO speed, we'll see eventually when it's released. 

> I need wifi to send the data to display.

The latest Pico W has wireless capability (not with the 5GHz band but 2.4GHz), perhaps you can create a DVI-sock for it as the one I linked before only seems to fit version 1 of the Pico (the one without WiFi-chip) from what I can tell, I might be wrong. 
Edit: It fits but you cannot solder the three pads in the front (the debug pads) since the Pico W doesn't have that same design. A bit worse mechanical strength for the DVI-sock then but probably not a big deal.

> I did find a chip that does a pcie interface with simple parallel bit input,

Regarding PCIe, that is a different ballgame. Now you are talking power hungry graphic cards, +12V needed and you would also need at least 33 MHz speed on the bus. There is no way near sufficient number of GPIOs on the ESP32-S3 to connect to PCI even if it had the GPIO speed (which I don't think it has, not as standard anyway). You can just count the number of pins needed to hook up the Databus, the Address pins and all the control signals on that datasheet you linked and you see it's way more than any small microcontroller has. In that case make your life easy and just use a RPi Zero 2 W would be my recommendation. It has a Mini-HDMI output too.

**Comment by BrainBoardz** [comment]
I would agree that if you really need HDMI on a budget the  RPi Zero 2 W or even a more full featured PI (e.g., Pi 4 or 5) would likely be the best way to go. There's even signage solutions that run on PIs. I have been able to get my ESP32S3 VGA board (based on Bitluni's design) to work fine with cheap VGA to HDMI converters. My board even supports audio to HDMI over i2s (with the right VGA converter). But, complete solutions are  more than just what the hardware is capable of. You ideally want full software support as well and decent performance. The ESP S3 VGA library as it stand is impressive, but it does struggle above 320X240 or so if you start to add other activities in the mix.  If you added the cost of the bits to convert the VGA signal to HDMI, for example, you would be spending very nearly what a Zero 2W costs (under $30 CAD), perhaps even more. The Pico can deliver DVI via an HDMI connector (so HDMI in a fashion), but again you have to overclock it to even get there. Not worth it IMO.  For signage scenarios I would go with something predictable, a Zero 2W or even a full on Pi if you want bells and whistles.
